src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] storage = (T[]) new Object[DEFAULT_CAPACITY];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        if (size == storage.length) {
+            T[] newStorage = (T[]) new Object[storage.length + (storage.length >> 1)];
+            for (int i = 0; i < storage.length; i++) {
+                newStorage[i] = storage[i];
+            }
+            storage = newStorage;
+        }
+        storage[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of array size");
+        }
+
+        if (index == size) {
+            add(value);
+            return;
+        }
 
+        if (size == storage.length) {
+            T[] newStorage = (T[]) new Object[storage.length + (storage.length >> 1)];
+            for (int i = 0; i < index; i++) {
+                newStorage[i] = storage[i];
+            }
+            newStorage[index] = value;
+            for (int i = index; i < storage.length; i++) {
+                newStorage[i + 1] = storage[i];
+            }
+            storage = newStorage;
+        } else {
+            for (int i = size - 1; i >= index; i--) {
+                storage[i + 1] = storage[i];
+            }
+            storage[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            return;
+        }
 
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of array size");
+        }
+        return storage[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of array size");
+        }
 
+        storage[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of array size");
+        }
+        T value = storage[index];
+        for (int i = index + 1; i < size; i++) {
+            storage[i - 1] = storage[i];
+        }
```

Comment:
Use System.arraycopy here
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html#use-systemarraycopy-to-move-your-array-elements